### PR TITLE
Added Rewrite MPEG-TS SI settings to the MPEG-TS SPAWN profile panel.

### PR DESCRIPTION
**Pull request by Jumper78**
Update profile.c
Added Rewrite MPEG-TS SI settings to the MPEG-TS SPAWN profile panel.

Currently, only the MPEG-TS profile has the expert options for MPEG-TS SI settings. Unfortunately, I need these settings for the TV stations I am receiving. Up to now, I was not able to use MPEG-TS SPAWN for recoding via ffmpeg. After adding the MPEG-TS SI Table(s) Settings panel to the expert options of MPEG-TS SPAWN and deselecting some of the rewrite options, that is now possible.

### CODE REVIEW

File: src/profile.c [hunk 1]
```diff
@@ -1512,8 +1512,77 @@ typedef struct profile_mpegts_spawn {
...
 } profile_mpegts_spawn_t;
 
+static int
+profile_spawn_rewrite_sid_set (void *in, const void *v)
+{
+  profile_mpegts_spawn_t *pro = (profile_mpegts_spawn_t *)in;
+  const uint16_t *val = v;
+  if (*val != pro->pro_rewrite_sid) {
+    if (*val > 0) {
+      pro->pro_rewrite_pmt =
```
Comment (Flole998) [inline src/profile.c:1530]: Please don't do this. Specify the value on each line, that makes it easier to read and less likely to break when changing this code later on.

Comment (Jumper78) [reply src/profile.c]: Understood! I also changed it at the position, where I copied it from. It is now fixed in the pass profile and in the spawn profile.

**Comment by Flole998** [comment]
I just thought about this and I am not sure if I understand correctly why this is necessary: Usually when you use the spawn profile you should be able to make all these settings on the spawned process so a stream with the correct settings arrives at Tvheadend? Why is that not working in this case?

**Comment by Jumper78** [comment]
> Why is that not working in this case?

With my TV stations, when having "Rewrite PMT" and "Rewrite NIT" set to 1, I end up loosing the video stream and end up just with an audio stream. This is the case for the pass filter. When then using the spawn filter, I can configure ffmpeg however I want, I will never be able to get the filtered out video stream back.
Or, if you look at it from the source code side: the parameters c.u.pass.m_rewrite_sid, c.u.pass.m_rewrite_pat, c.u.pass.m_rewrite_pmt, c.u.pass.m_rewrite_sdt, c.u.pass.m_rewrite_nit and c.u.pass.m_rewrite_eit are set to 1 in the background without me having a chance to influence on that.

Maybe I have still not understood everything about the stream I am transcoding, but the fact that I am now able to change these options is working like a charm for me.

**Comment by Flole998** [comment]
What TV stations are you using and what Player is not showing the video stream? I really appreciate your PR and the work you did, but I want to make sure we aren't "covering" another issue here by just throwing more options at it. Maybe I can have a look at what's going in and what's coming out to see why the rewriting is necessary.

**Comment by Jumper78** [comment]
@Flole998: Chances are high that my PR and my settings are just a workaround. I appreciate that you take some time to look at my problem.

Instead of a player, I use ffmpeg (ffmpeg version 4.1.6-1~deb10u1+rpt1 on RasPi4) at different positions of the stream to analyse the content. I am looking at terrestrial TV streams in South Korea (with an Hauppauge WinTV-dualHD for ATSC).

1. stream of the station MBC _mux_ with
    - MPEG-TS Pass-thru/built-in as default stream profile
    - Rewrite Service ID = 1, Rewrite PMT = 1, Rewrite PAT = 1, Rewrite SDT = 1, Rewrite NIT = 1, Rewrite EIT = 1

```
[mpeg2video @ 0x1719820] Invalid frame dimensions 0x0.
    Last message repeated 7 times
[mpegts @ 0x170e170] Could not find codec parameters for stream 3 (Unknown: none ([5][0][0][0] / 0x0005)): unknown codec
Consider increasing the value for the 'analyzeduration' and 'probesize' options
[mpegts @ 0x170e170] Could not find codec parameters for stream 4 (Unknown: none ([11][0][0][0] / 0x000B)): unknown codec
Consider increasing the value for the 'analyzeduration' and 'probesize' options
Input #0, mpegts, from 'http://<user:pass>@<ip>:9981/play/ticket/stream/mux/c967e5c90c2581b53f15c4577cf1171d':
  Duration: N/A, start: 8614.151589, bitrate: N/A
  Program 1 
    Stream #0:0[0x11]: Video: mpeg2video (Main) ([2][0][0][0] / 0x0002), yuv420p(tv, top first), 1920x1080 [SAR 1:1 DAR 16:9], Closed Captions, 29.97 fps, 29.97 tbr, 90k tbn, 59.94 tbc
    Stream #0:1[0x14](kor): Audio: ac3 ([129][0][0][0] / 0x0081), 48000 Hz, stereo, fltp, 448 kb/s
    Stream #0:2[0x15](kor): Audio: ac3 ([129][0][0][0] / 0x0081), 48000 Hz, stereo, fltp, 192 kb/s (visual impaired) (descriptions)
    Stream #0:3[0x1f]: Unknown: none ([5][0][0][0] / 0x0005)
    Stream #0:4[0x200]: Unknown: none ([11][0][0][0] / 0x000B)
```

2. stream of the MBC _service_ with
    - MPEG-TS Pass-thru/built-in as default stream profile
    - _Rewrite Service ID = 1_, _Rewrite PMT = 1_, Rewrite PAT = 1, Rewrite SDT = 1, _Rewrite NIT = 1_, Rewrite EIT = 1

```
Input #0, mpegts, from 'http://<user:pass>@<ip>:9981/play/ticket/stream/service/f7f049a4b10db70babc875375cab20a2':
  Duration: N/A, start: 9693.543589, bitrate: 448 kb/s
  Program 1 
  No Program
    Stream #0:0[0x14]: Audio: ac3, 48000 Hz, stereo, fltp, 448 kb/s`
```

3. stream of the MBC _service_ with
    - MPEG-TS Pass-thru/built-in as default stream profile
    - _Rewrite Service ID = 0_, _Rewrite PMT = 0_, Rewrite PAT = 1, Rewrite SDT = 1, _Rewrite NIT = 0_, Rewrite EIT = 1

```
[mpeg2video @ 0x1fef520] Invalid frame dimensions 0x0.
    Last message repeated 13 times
[mpegts @ 0x1fe3170] Could not find codec parameters for stream 2 (Unknown: none ([5][0][0][0] / 0x0005)): unknown codec
Consider increasing the value for the 'analyzeduration' and 'probesize' options
[mpegts @ 0x1fe3170] Could not find codec parameters for stream 3 (Unknown: none ([11][0][0][0] / 0x000B)): unknown codec
Consider increasing the value for the 'analyzeduration' and 'probesize' options
Input #0, mpegts, from 'http://<user:pass>@<ip>:9981/play/ticket/stream/service/f7f049a4b10db70babc875375cab20a2':
  Duration: N/A, start: 9916.263589, bitrate: N/A
  Program 1 
    Stream #0:0[0x11]: Video: mpeg2video (Main) ([2][0][0][0] / 0x0002), yuv420p(tv, top first), 1920x1080 [SAR 1:1 DAR 16:9], Closed Captions, 29.97 fps, 29.97 tbr, 90k tbn, 59.94 tbc
    Stream #0:1[0x14](kor): Audio: ac3 ([129][0][0][0] / 0x0081), 48000 Hz, stereo, fltp, 448 kb/s
    Stream #0:2[0x1f]: Unknown: none ([5][0][0][0] / 0x0005)
    Stream #0:3[0x200]: Unknown: none ([11][0][0][0] / 0x000B)`
```

So, with the activated rewrite options, TVHeadend kills all streams except the audio stream(s). Some other TV stations (EBS1, SBS) do also suffer from the problem while others (KBS1, KBS2, EBS2) do not have the problem. Unfortunately, there is very little information about the composition of the streams (at least in the languages I speak well enough). Any suggestion for any further analysis?

**Comment by Saentist** [comment]
@Jumper78 can you analyze with https://github.com/tsduck/tsduck
similar problem with pass profile https://tvheadend.org/issues/5062

**Comment by Jumper78** [comment]
@Saentist : TSDuck is mighty tool! It might even help me with getting EPG running for those channels. But it also took me a while to getting something useful out of it.
Anyway, I analysed the PMT tables of the streams. I attach the output of one of the faulty streams at the end of this post. Interestingly, all of the affected streams use the same PID: 0x0010, the others use 0x0020.

| name      | KBS1        | KBS2        | MBC         | SBS         | EBS1        | EBS2        |
| ----------|-------------|-------------|-------------|-------------|-------------|-------------|
| TID       | 2 (0x02)    | 2 (0x02)    | 2 (0x02)    | 2 (0x02)    | 2 (0x02)    | 2 (0x02)    |
| PID       | 32 (0x0020) | 32 (0x0020) | 16 (0x0010) | 16 (0x0010) | 16 (0x0010) | 32 (0x0020) |
| affected? | no          | no          | yes         | yes         | yes         | no          |

My current hypothesis: DVB_NIT_PID is also set to 0x10. In src/muxer/muxer_pass.c, in line 635 a PID, the NIT-rewrite flag and DVB_NIT_PID are all brought together in an `if`. Maybe that is the position where my stream is kicked out. What I do not understand: if the hypothesis is true, why is the NIT-rewrite then affecting only the video stream and not the audio stream?
I will try out a couple of things tonight and will hopefully find a fix for the real problem.

Here the PMT section from MBS (recording with `tsp -I dvb --frequency <frequency> -P until --seconds 60 -O file <filename>.ts`, analysing with `tspsi <filename>.ts`:
```
* PMT, TID 2 (0x02), PID 16 (0x0010)
  Version: 7, sections: 1, total size: 121 bytes
  - Section 0:
    Program: 1 (0x0001), PCR PID: 17 (0x0011)
    Elementary stream: type 0x02 (MPEG-2 Video), PID: 17 (0x0011)
    Elementary stream: type 0x81 (ATSC AC-3 Audio), PID: 20 (0x0014)
    - Descriptor 0: ISO-639 Language (0x0A, 10), 4 bytes
      Language: kor, Type: 0x00 (undefined)
    - Descriptor 1: ATSC AC-3 Audio (0x81, 129), 6 bytes
      Sample rate: 48 kHz (0x0)
      AC-3 coding version: 0x08 (8)
      Bit rate: 384 kb/s (0x0E) max
      Surround mode: Not indicated (0x0)
      Bitstream mode: main audio service: complete main (CM) (0x0)
      Num. channels: 6 max (0xD)
      Full service: true
      Main audio service id: 0
      Priority: 3
      Text: ""
    Elementary stream: type 0x05 (MPEG-2 Private sections), PID: 31 (0x001F)
    - Descriptor 0: DSM-CC Association Tag (0x14, 20), 13 bytes
      Association tag: 0x0081 (129), use: 0x0000 (0)
      Selector bytes (8 bytes): FF FF FF FF FF FF FF FF
    - Descriptor 1: Stream Identifier (0x52, 82), 1 bytes
      Component tag: 129 (0x81)
    - Descriptor 2: Application Signalling (0x6F, 111), 3 bytes
      Application type: 6 (0x0006), AIT Version: 2 (0x02)
    Elementary stream: type 0x0B (DSM-CC U-N), PID: 512 (0x0200)
    - Descriptor 0: DSM-CC Association Tag (0x14, 20), 13 bytes
      Association tag: 0x0002 (2), use: 0x0000 (0)
      Selector bytes (8 bytes): FF FF FF FF FF FF FF FF
    - Descriptor 1: Stream Identifier (0x52, 82), 1 bytes
      Component tag: 2 (0x02)
    - Descriptor 2: Data Broadcast Id (0x66, 102), 4 bytes
      Data broadcast id: 0x00F0 (240, MHP Object Carousel)
      Data Broadcast selector (2 bytes): 00 06
    - Descriptor 3: DSM-CC Carousel Identifier (0x13, 19), 22 bytes
      Carousel id: 0x80000001 (2147483649)
      Private data (18 bytes):
        0000:  01 00 00 00 0F E2 00 00 A5 D0 00 00 00 A5 D0 FF  ................
        0010:  01 00                                            ..
```

**Comment by perexg** [comment]
@Flole998 : I think that it should be merged. Appearently, this MPEG-TS stream is not ETSI EN 300 468 conformant:

etsi1

**Comment by Flole998** [comment]
The PMT is broadcasted on the PID that is supposed to be used for the NIT, correct? We could probably detect that and then disable the rewriting automaticaly so we don't end up with broken streams? 

As @Jumper78 wants to do some further investigation I will wait until that is done, and as @perexg has given his okay I will merge it then if no better solution is found :)

**Comment by perexg** [comment]
> The PMT is broadcasted on the PID that is supposed to be used for the NIT, correct? We could probably detect that and then disable the rewriting automaticaly so we don't end up with broken streams?

It may be a good idea. If PMT PID is equal to any predefined SDT / NIT / EIT PID, the rewrite should be turned automatically off for the corresponding PID.

Basically, the PMT parser _must be_ called in `pass_muxer_write_ts()` - `dvb_table_parse(&pm->pm_pmt, "-", tsb, l, 1, 0, pass_muxer_pmt_cb)`.

**Comment by Flole998** [comment]
How about adding something like `pid != pm->pm_pmt_pid` to each condition (except for the PMT one, that one would need a pid != SDT && pid != NIT....) in line 632 there? That should be exactly what we need?

I think we should still merge this PR, but it would be great to get the other fix aswell.

**Comment by perexg** [comment]
> How about adding something like `pid != pm->pm_pmt_pid` to each condition (except for the PMT one, that one would need a pid != SDT && pid != NIT....) in line 632 there? That should be exactly what we need?

It's better to set directly the pm_rewrite variables in `pass_muxer_reconfigure`.  The mentioned conditions are evaluated for each TS packet. The pm_rewrite variables are local and set when the stream changes.

Also note that we have already such "xxx PID shared with A/V, rewrite disabled" conditions there :-) So it's just about to tune this code also for this MPEG-TS stream.

**Comment by Jumper78** [comment]
My hypothesis seems to be correct. I changed `DVB_NIT_PID` from `0x10`to `0x0F`and tried again to grab with ffmpeg.

4. Repeating step 3 from my last test to make sure nothing else has changed: stream of the MBC service with
    - MPEG-TS Pass-thru/built-in as default stream profile
    - Rewrite Service ID = 0, Rewrite PMT = 0, Rewrite PAT = 1, Rewrite SDT = 1, Rewrite NIT = 0, Rewrite EIT = 1

```
Input #0, mpegts, from 'http://<user:pass>@<ip>:9981/play/ticket/stream/service/f7f049a4b10db70babc875375cab20a2':
  Duration: N/A, start: 91944.577900, bitrate: N/A
  Program 1 
    Stream #0:0[0x11]: Video: mpeg2video (Main) ([2][0][0][0] / 0x0002), yuv420p(tv, top first), 1920x1080 [SAR 1:1 DAR 16:9], Closed Captions, 29.97 fps, 29.97 tbr, 90k tbn, 59.94 tbc
    Stream #0:1[0x14](kor): Audio: ac3 ([129][0][0][0] / 0x0081), 48000 Hz, stereo, fltp, 448 kb/s
    Stream #0:2[0x15](kor): Audio: ac3 ([129][0][0][0] / 0x0081), 48000 Hz, stereo, fltp, 192 kb/s (visual impaired) (descriptions)
    Stream #0:3[0x1f]: Unknown: none ([5][0][0][0] / 0x0005)
    Stream #0:4[0x200]: Unknown: none ([11][0][0][0] / 0x000B)
```

5. Repeating step 2 from my last test but with `DVB_NIT_PID` set to `0x0F`: stream of the MBC service with
    - MPEG-TS Pass-thru/built-in as default stream profile
    - Rewrite Service ID = 1, Rewrite PMT = 1, Rewrite PAT = 1, Rewrite SDT = 1, Rewrite NIT = 1, Rewrite EIT = 1

```
Input #0, mpegts, from 'http://<user:pass>@<ip>:9981/play/ticket/stream/service/f7f049a4b10db70babc875375cab20a2':
  Duration: N/A, start: 92077.761900, bitrate: N/A
  Program 1 
    Stream #0:0[0x11]: Video: mpeg2video (Main) ([2][0][0][0] / 0x0002), yuv420p(tv, top first), 1920x1080 [SAR 1:1 DAR 16:9], Closed Captions, 29.97 fps, 29.97 tbr, 90k tbn, 59.94 tbc
    Stream #0:1[0x14](kor): Audio: ac3 ([129][0][0][0] / 0x0081), 48000 Hz, stereo, fltp, 448 kb/s
    Stream #0:2[0x15](kor): Audio: ac3 ([129][0][0][0] / 0x0081), 48000 Hz, stereo, fltp, 192 kb/s (visual impaired) (descriptions)
```

Conclusion: the video stream disappears because its PID in the PMT tables conflicts with `DVB_NIT_PID`. My current PR is a workaround. The fixes for @perexg and @Flole998 also sound promising.

**Comment by Jumper78** [comment]
> It's better to set directly the pm_rewrite variables in `pass_muxer_reconfigure`. The mentioned conditions are evaluated for each TS packet. The pm_rewrite variables are local (to the pass muxer instance) and set when the stream changes.
> 
> Also note that we have already such "xxx PID shared with A/V, rewrite disabled" conditions there :-) So it's just about to tune this code also for this MPEG-TS stream.

@perexg I have written a fix that works for me. I tried to follow your guidance from this comment. Please review #1402 and let me know if you have any suggestions for improvements.

**Comment by Flole998** [comment]
Looks good to me, I would have split this into to PRs and merged this one aswell but it's also fine in one PR. Please squash your changes, then I will merge unless @perexg sees anything wrong with it.

**Comment by Flole998** [comment]
Is there any reason why you didn't include the PAT? I think that one should be checked aswell?

**Comment by Jumper78** [comment]
> Is there any reason why you didn't include the PAT? I think that one should be checked aswell?

My understanding is that PAT always has to be on PID 0x00. Please correct me if that assumption is wrong. If that assumption is correct, there will never be the case that the PAT PID conflicts with SDT, NIT or EIT PID.

**Comment by Flole998** [comment]
But couldn't it conflict with the PMT, just like the PMT could conflict with the SDT/NIT/EIT?

**Comment by Jumper78** [comment]
> But couldn't it conflict with the PMT, just like the PMT could conflict with the SDT/NIT/EIT?

My understanding is that PAT has to have PID 0x00 and no other stream is allowed to have the same PID. So, if there is a conflict between PAT and SDT/NIT/EIT, there would be something seriously wrong with the stream.

**Pull request merged**